File: go/src/model_router.go
// Model Router — Go
// Unified multi-provider chat-completion client for Claude, DeepSeek, and OpenAI.
//
// go get github.com/anthropics/anthropic-sdk-go@v1.4.0
// export MODEL_PROVIDER=anthropic|deepseek|openai
// export ANTHROPIC_API_KEY / DEEPSEEK_API_KEY / OPENAI_API_KEY

package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

// ── Types ──────────────────────────────────────────────────────────────────────
type ContentBlockType string

const (
	BlockText    ContentBlockType = "text"
	BlockToolUse ContentBlockType = "tool_use"
)

type TextBlock struct {
	Type ContentBlockType `json:"type"`
	Text string           `json:"text"`
}

type ToolUseBlock struct {
	Type  ContentBlockType       `json:"type"`
	ID    string                 `json:"id"`
	Name  string                 `json:"name"`
	Input map[string]interface{} `json:"input"`
}

type ContentBlock struct {
	Type  ContentBlockType       `json:"type"`
	Text  string                 `json:"text,omitempty"`
	ID    string                 `json:"id,omitempty"`
	Name  string                 `json:"name,omitempty"`
	Input map[string]interface{} `json:"input,omitempty"`
}

type Usage struct {
	InputTokens  int `json:"input_tokens"`
	OutputTokens int `json:"output_tokens"`
}

type UnifiedResponse struct {
	Content    []ContentBlock
	StopReason string
	Usage      Usage
}

func (r *UnifiedResponse) ToAssistantMessage() map[string]interface{} {
	blocks := make([]interface{}, 0, len(r.Content))
	for _, b := range r.Content {
		if b.Type == BlockText {
			blocks = append(blocks, map[string]interface{}{"type": "text", "text": b.Text})
		} else if b.Type == BlockToolUse {
			blocks = append(blocks, map[string]interface{}{"type": "tool_use", "id": b.ID, "name": b.Name, "input": b.Input})
		}
	}
	return map[string]interface{}{"role": "assistant", "content": blocks}
}

// ── Default models ─────────────────────────────────────────────────────────────
var defaultModels = map[string]string{
	"anthropic":       "claude-opus-4-6",
	"deepseek":        "deepseek-v4-pro",
	"openai":          "gpt-5.5",
}

// ── Model Router ───────────────────────────────────────────────────────────────
type ModelRouter struct {
	Provider string
	Model    string
	baseURL  string
	apiKey   string
	client   *http.Client
}

func NewModelRouter(provider, model string) (*ModelRouter, error) {
	if provider == "" {
		provider = os.Getenv("MODEL_PROVIDER")
		if provider == "" {
			provider = "anthropic"
		}
	}
	if model == "" {
		model = os.Getenv("MODEL_NAME")
		if model == "" {
			var ok bool
			if model, ok = defaultModels[provider]; !ok {
				return nil, fmt.Errorf("unknown MODEL_PROVIDER '%s'", provider)
			}
		}
	}
	var baseURL, apiKey string
	switch provider {
	case "anthropic":
		baseURL = "https://api.anthropic.com"
		apiKey  = os.Getenv("ANTHROPIC_API_KEY")
	case "deepseek":
		baseURL = "https://api.deepseek.com/anthropic"
		apiKey  = os.Getenv("DEEPSEEK_API_KEY")
	case "openai":
		baseURL = "https://api.openai.com"
		apiKey  = os.Getenv("OPENAI_API_KEY")
	default:
		return nil, fmt.Errorf("unknown MODEL_PROVIDER '%s'", provider)
	}
	return &ModelRouter{
		Provider: provider,
		Model:    model,
		baseURL:  baseURL,
		apiKey:   apiKey,
		client:   &http.Client{Timeout: 90 * time.Second},
	}, nil
}

type CreateOptions struct {
	Messages  []map[string]interface{}
	Tools     []map[string]interface{}
	System    string
	MaxTokens int
}

func (r *ModelRouter) Create(ctx context.Context, opts CreateOptions) (*UnifiedResponse, error) {
	if opts.MaxTokens == 0 {
		opts.MaxTokens = 4096
	}
	switch r.Provider {
	case "anthropic", "deepseek":
		return r.createAnthropic(ctx, opts)
	case "openai":
		return r.createOpenAI(ctx, opts)
	}
	return nil, fmt.Errorf("unsupported provider: %s", r.Provider)
}

// ── Anthropic-format request ───────────────────────────────────────────────────
type anthropicRequest struct {
	Model     string                   `json:"model"`
	MaxTokens int                      `json:"max_tokens"`
	Messages  []map[string]interface{} `json:"messages"`
	System    string                   `json:"system,omitempty"`
	Tools     []map[string]interface{} `json:"tools,omitempty"`
}

type anthropicResponse struct {
	Content    []ContentBlock `json:"content"`
	StopReason string         `json:"stop_reason"`
	Usage      Usage          `json:"usage"`
}

func (r *ModelRouter) createAnthropic(ctx context.Context, opts CreateOptions) (*UnifiedResponse, error) {
	body := anthropicRequest{
		Model:     r.Model,
		MaxTokens: opts.MaxTokens,
		Messages:  opts.Messages,
		System:    opts.System,
		Tools:     opts.Tools,
	}
	data, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, r.baseURL+"/v1/messages", bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", r.apiKey)
	req.Header.Set("anthropic-version", "2023-06-01")

	resp, err := r.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	respData, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("anthropic API %d: %s", resp.StatusCode, string(respData))
	}
	var ar anthropicResponse
	if err := json.Unmarshal(respData, &ar); err != nil {
		return nil, err
	}
	return &UnifiedResponse{Content: ar.Content, StopReason: ar.StopReason, Usage: ar.Usage}, nil
}

// ── OpenAI-format request ──────────────────────────────────────────────────────
type openAIMessage struct {
	Role      string      `json:"role"`
	Content   interface{} `json:"content"`
	ToolCalls interface{} `json:"tool_calls,omitempty"`
}

type openAIRequest struct {
	Model     string                   `json:"model"`
	MaxTokens int                      `json:"max_tokens"`
	Messages  []openAIMessage          `json:"messages"`
	Tools     []map[string]interface{} `json:"tools,omitempty"`
}

func (r *ModelRouter) createOpenAI(ctx context.Context, opts CreateOptions) (*UnifiedResponse, error) {
	var messages []openAIMessage
	if opts.System != "" {
		messages = append(messages, openAIMessage{Role: "system", Content: opts.System})
	}
	for _, m := range opts.Messages {
		messages = append(messages, openAIMessage{Role: m["role"].(string), Content: m["content"]})
	}
	var oaTools []map[string]interface{}
	for _, t := range opts.Tools {
		oaTools = append(oaTools, map[string]interface{}{
			"type": "function",
			"function": map[string]interface{}{
				"name":        t["name"],
				"description": t["description"],
				"parameters":  t["input_schema"],
			},
		})
	}
	body := openAIRequest{Model: r.Model, MaxTokens: opts.MaxTokens, Messages: messages, Tools: oaTools}
	data, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, r.baseURL+"/v1/chat/completions", bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+r.apiKey)

	resp, err := r.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	respData, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("openai API %d: %s", resp.StatusCode, string(respData))
	}
	var result struct {
		Choices []struct {
			Message struct {
				Content   string `json:"content"`
				ToolCalls []struct {
					ID       string `json:"id"`
					Function struct {
						Name      string `json:"name"`
						Arguments string `json:"arguments"`
					} `json:"function"`
				} `json:"tool_calls"`
			} `json:"message"`
		} `json:"choices"`
		Usage struct {
			PromptTokens     int `json:"prompt_tokens"`
			CompletionTokens int `json:"completion_tokens"`
		} `json:"usage"`
	}
	if err := json.Unmarshal(respData, &result); err != nil {
		return nil, err
	}
	if len(result.Choices) == 0 {
		return nil, fmt.Errorf("no choices in OpenAI response")
	}
	choice := result.Choices[0]
	var content []ContentBlock
	if choice.Message.Content != "" {
		content = append(content, ContentBlock{Type: BlockText, Text: choice.Message.Content})
	}
	stopReason := "end_turn"
	for _, tc := range choice.Message.ToolCalls {
		var input map[string]interface{}
		_ = json.Unmarshal([]byte(tc.Function.Arguments), &input)
		content = append(content, ContentBlock{Type: BlockToolUse, ID: tc.ID, Name: tc.Function.Name, Input: input})
		stopReason = "tool_use"
	}
	return &UnifiedResponse{
		Content:    content,
		StopReason: stopReason,
		Usage:      Usage{InputTokens: result.Usage.PromptTokens, OutputTokens: result.Usage.CompletionTokens},
	}, nil
}

// ── Tool result message ────────────────────────────────────────────────────────
func MakeToolResultMessage(toolUseID, content string, isError bool) map[string]interface{} {
	block := map[string]interface{}{
		"type":        "tool_result",
		"tool_use_id": toolUseID,
		"content":     content,
	}
	if isError {
		block["is_error"] = true
	}
	return map[string]interface{}{"role": "user", "content": []interface{}{block}}
}

File: java/src/ModelRouter.java
/**
 * Model Router — Java
 * Unified multi-provider chat-completion client for Claude, DeepSeek, and OpenAI.
 * Uses OkHttp + Jackson. Algorithms pinned, no reflection-based deserialization exploits.
 *
 * export MODEL_PROVIDER=anthropic|deepseek|openai
 * export ANTHROPIC_API_KEY / DEEPSEEK_API_KEY / OPENAI_API_KEY
 */

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.*;
import okhttp3.*;

import java.io.IOException;
import java.util.*;
import java.util.concurrent.TimeUnit;

public class ModelRouter {

    // ── Config ─────────────────────────────────────────────────────────────────
    static final Map<String, String> DEFAULT_MODELS = Map.of(
        "anthropic", "claude-opus-4-6",
        "deepseek",  "deepseek-v4-pro",
        "openai",    "gpt-5.5"
    );

    static final Map<String, String> BASE_URLS = Map.of(
        "anthropic", "https://api.anthropic.com",
        "deepseek",  "https://api.deepseek.com/anthropic",
        "openai",    "https://api.openai.com"
    );

    static final ObjectMapper JSON = new ObjectMapper();
    static final MediaType JSON_TYPE = MediaType.get("application/json; charset=utf-8");
    static final OkHttpClient HTTP = new OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(90, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build();

    // ── Types ──────────────────────────────────────────────────────────────────
    public enum BlockType { TEXT, TOOL_USE }

    public static class ContentBlock {
        public final BlockType type;
        public final String text;
        public final String id;
        public final String name;
        public final Map<String, Object> input;

        private ContentBlock(BlockType type, String text, String id, String name, Map<String, Object> input) {
            this.type = type; this.text = text; this.id = id; this.name = name; this.input = input;
        }

        public static ContentBlock text(String text) {
            return new ContentBlock(BlockType.TEXT, text, null, null, null);
        }

        public static ContentBlock toolUse(String id, String name, Map<String, Object> input) {
            return new ContentBlock(BlockType.TOOL_USE, null, id, name, input);
        }

        public Map<String, Object> toAssistantBlock() {
            Map<String, Object> m = new LinkedHashMap<>();
            if (type == BlockType.TEXT) {
                m.put("type", "text"); m.put("text", text);
            } else {
                m.put("type", "tool_use"); m.put("id", id); m.put("name", name); m.put("input", input);
            }
            return m;
        }
    }

    public static class UnifiedResponse {
        public final List<ContentBlock> content;
        public final String stopReason;
        public final int inputTokens;
        public final int outputTokens;

        UnifiedResponse(List<ContentBlock> content, String stopReason, int inputTokens, int outputTokens) {
            this.content = content; this.stopReason = stopReason;
            this.inputTokens = inputTokens; this.outputTokens = outputTokens;
        }

        public Map<String, Object> toAssistantMessage() {
            List<Map<String, Object>> blocks = new ArrayList<>();
            for (ContentBlock b : content) blocks.add(b.toAssistantBlock());
            return Map.of("role", "assistant", "content", blocks);
        }
    }

    // ── Instance ───────────────────────────────────────────────────────────────
    public final String provider;
    public final String model;
    private final String baseUrl;
    private final String apiKey;

    public ModelRouter() {
        this(null, null);
    }

    public ModelRouter(String provider, String model) {
        this.provider = Optional.ofNullable(provider)
            .or(() -> Optional.ofNullable(System.getenv("MODEL_PROVIDER")))
            .orElse("anthropic");

        if (!DEFAULT_MODELS.containsKey(this.provider)) {
            throw new IllegalArgumentException("Unknown MODEL_PROVIDER: " + this.provider);
        }
        this.model = Optional.ofNullable(model)
            .or(() -> Optional.ofNullable(System.getenv("MODEL_NAME")))
            .orElse(DEFAULT_MODELS.get(this.provider));
        this.baseUrl = BASE_URLS.get(this.provider);

        String key = switch (this.provider) {
            case "openai"    -> System.getenv("OPENAI_API_KEY");
            case "deepseek"  -> System.getenv("DEEPSEEK_API_KEY");
            default          -> System.getenv("ANTHROPIC_API_KEY");
        };
        if (key == null || key.isBlank()) throw new IllegalStateException("API key not set for provider: " + this.provider);
        this.apiKey = key;
    }

    // ── Main create() ──────────────────────────────────────────────────────────
    public UnifiedResponse create(
        List<Map<String, Object>> messages,
        List<Map<String, Object>> tools,
        String system,
        int maxTokens
    ) throws Exception {
        if (maxTokens <= 0) maxTokens = 4096;
        return switch (provider) {
            case "anthropic", "deepseek" -> createAnthropic(messages, tools, system, maxTokens);
            case "openai"                -> createOpenAI(messages, tools, system, maxTokens);
            default -> throw new IllegalStateException("Unsupported provider: " + provider);
        };
    }

    // ── Anthropic-format ───────────────────────────────────────────────────────
    private UnifiedResponse createAnthropic(
        List<Map<String, Object>> messages,
        List<Map<String, Object>> tools,
        String system, int maxTokens
    ) throws Exception {
        ObjectNode body = JSON.createObjectNode();
        body.put("model", model);
        body.put("max_tokens", maxTokens);
        body.set("messages", JSON.valueToTree(messages));
        if (system != null && !system.isBlank()) body.put("system", system);
        if (tools != null && !tools.isEmpty()) body.set("tools", JSON.valueToTree(tools));

        Request req = new Request.Builder()
            .url(baseUrl + "/v1/messages")
            .header("x-api-key", apiKey)
            .header("anthropic-version", "2023-06-01")
            .post(RequestBody.create(JSON.writeValueAsBytes(body), JSON_TYPE))
            .build();

        return executeAndParseAnthropic(req);
    }

    private UnifiedResponse executeAndParseAnthropic(Request req) throws Exception {
        try (Response resp = HTTP.newCall(req).execute()) {
            String bodyStr = resp.body().string();
            if (!resp.isSuccessful()) throw new IOException("Anthropic API " + resp.code() + ": " + bodyStr);
            JsonNode root = JSON.readTree(bodyStr);
            String stopReason = root.path("stop_reason").asText("end_turn");
            int inputTokens   = root.path("usage").path("input_tokens").asInt(0);
            int outputTokens  = root.path("usage").path("output_tokens").asInt(0);
            List<ContentBlock> content = new ArrayList<>();
            for (JsonNode b : root.path("content")) {
                String type = b.path("type").asText();
                if ("text".equals(type)) {
                    content.add(ContentBlock.text(b.path("text").asText()));
                } else if ("tool_use".equals(type)) {
                    Map<String, Object> input = JSON.convertValue(b.path("input"), Map.class);
                    content.add(ContentBlock.toolUse(b.path("id").asText(), b.path("name").asText(), input));
                }
            }
            return new UnifiedResponse(content, stopReason, inputTokens, outputTokens);
        }
    }

    // ── OpenAI-format ──────────────────────────────────────────────────────────
    private UnifiedResponse createOpenAI(
        List<Map<String, Object>> messages,
        List<Map<String, Object>> tools,
        String system, int maxTokens
    ) throws Exception {
        List<Map<String, Object>> oaMessages = new ArrayList<>();
        if (system != null && !system.isBlank()) oaMessages.add(Map.of("role", "system", "content", system));
        oaMessages.addAll(messages);

        ObjectNode body = JSON.createObjectNode();
        body.put("model", model);
        body.put("max_tokens", maxTokens);
        body.set("messages", JSON.valueToTree(oaMessages));

        if (tools != null && !tools.isEmpty()) {
            List<Map<String, Object>> oaTools = new ArrayList<>();
            for (Map<String, Object> t : tools) {
                oaTools.add(Map.of("type", "function", "function", Map.of(
                    "name",        t.get("name"),
                    "description", t.getOrDefault("description", ""),
                    "parameters",  t.getOrDefault("input_schema", Map.of("type", "object", "properties", Map.of()))
                )));
            }
            body.set("tools", JSON.valueToTree(oaTools));
            body.put("tool_choice", "auto");
        }

        Request req = new Request.Builder()
            .url(baseUrl + "/v1/chat/completions")
            .header("Authorization", "Bearer " + apiKey)
            .post(RequestBody.create(JSON.writeValueAsBytes(body), JSON_TYPE))
            .build();

        try (Response resp = HTTP.newCall(req).execute()) {
            String bodyStr = resp.body().string();
            if (!resp.isSuccessful()) throw new IOException("OpenAI API " + resp.code() + ": " + bodyStr);
            JsonNode root   = JSON.readTree(bodyStr);
            JsonNode choice = root.path("choices").path(0);
            JsonNode msg    = choice.path("message");
            int inputTokens  = root.path("usage").path("prompt_tokens").asInt(0);
            int outputTokens = root.path("usage").path("completion_tokens").asInt(0);

            List<ContentBlock> content = new ArrayList<>();
            String stopReason = "end_turn";
            String textContent = msg.path("content").asText(null);
            if (textContent != null && !textContent.isBlank()) content.add(ContentBlock.text(textContent));
            JsonNode toolCalls = msg.path("tool_calls");
            if (toolCalls.isArray() && toolCalls.size() > 0) {
                stopReason = "tool_use";
                for (JsonNode tc : toolCalls) {
                    String argsStr = tc.path("function").path("arguments").asText("{}");
                    Map<String, Object> input = JSON.readValue(argsStr, Map.class);
                    content.add(ContentBlock.toolUse(
                        tc.path("id").asText(),
                        tc.path("function").path("name").asText(),
                        input
                    ));
                }
            }
            return new UnifiedResponse(content, stopReason, inputTokens, outputTokens);
        }
    }

    // ── Tool result message ───────────────────────────────────────────────────
    public static Map<String, Object> makeToolResultMessage(String toolUseId, String content, boolean isError) {
        Map<String, Object> block = new LinkedHashMap<>();
        block.put("type", "tool_result");
        block.put("tool_use_id", toolUseId);
        block.put("content", content);
        if (isError) block.put("is_error", true);
        return Map.of("role", "user", "content", List.of(block));
    }
}
